File: tree/0437_path_sum_iii/0437_path_sum_iii.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, sum int) int {
	m := make(map[int]int)
	m[0] = 1
	ans := 0
	helper(root, sum, m, 0, &ans)
	return ans
}

func helper(root *TreeNode, target int, m map[int]int, cur int, ans *int) {
	fmt.Println("##### m is", m)
	if root == nil {
		return
	}
	cur = cur + root.Val
	*ans += m[cur-target]
	m[cur]++
	if root.Left != nil {
		helper(root.Left, target, m, cur, ans)
	}
	if root.Right != nil {
		helper(root.Right, target, m, cur, ans)
	}

	m[cur]--
}

func main() {
	root := &TreeNode{10, nil, nil}
	root.Left = &TreeNode{5, nil, nil}
	root.Right = &TreeNode{-3, nil, nil}
	root.Left.Left = &TreeNode{3, nil, nil}
	root.Left.Right = &TreeNode{2, nil, nil}
	root.Right.Right = &TreeNode{11, nil, nil}
	res := pathSum(root, 8)
	fmt.Println(res)
}

File: tree/0437_path_sum_iii/0437_path_sum_iii.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def pathSum(self, root, sum):
        def dfs(root, prevSum, sum):
            if not root:
                return
            currSum = prevSum + root.val
            if currSum - sum in rec:
                self.count += rec[currSum - sum]
            if currSum in rec:
                rec[currSum] += 1
            else:
                rec[currSum] =1
            dfs(root.left, currSum, sum)
            dfs(root.right, currSum, sum)
            rec[currSum] -= 1

        self.count = 0
        rec = {0:1}
        dfs(root, 0, sum)
        return self.count

root = TreeNode(10)
root.left = TreeNode(5)
root.right = TreeNode(-3)
root.left.left = TreeNode(3)
root.left.right = TreeNode(2)
root.right.right = TreeNode(11)
res = Solution().pathSum(root, 8)
print(res)